README.rst
RnsUtils
========

RnsUtils is Renoise related library and companion utility.
It is meant for developers aiming to use generate renoise file or renoise users wanting to convert SoundFont 2 or SFZ to renoise instruments,
reencode instruments to .flac or .ogg, manipulate tags and comments in instruments or organise a xrni library based on those tags.

Installation
------------

RnsUtils is installable from PyPI with a single pip command::

    pip install rnsutils

Alternatively, RnsUtils can be run directly from sources after a git pull::

    git clone https://gitlab.com/zeograd/rnsutils.git
    cd rnsutils && python setup.py install


sf2toxrni
---------

**sf2toxrni** is a command line utility which convert SoundFont 2 instruments into renoise instruments (.xrni).
It parses a SoundFont 2 file instrument list and generate one .xnri file for each instrument using as much information
from the SoundFont 2 instrument properties (generators) as possible. All read properties will be injected in a template
renoise instrument.

::

    usage: sf2toxrni [-h] [-d] [-e {none,flac,ogg}] [-f] [-q] [-u] [--no-unused]
                     [-o OUTPUT_DIR] [-t TEMPLATE]
                     sf2_filename [sf2_filename ...]

    GPL v3+ 2016-2017 Olivier Jolly

    positional arguments:
      sf2_filename          input file in SoundFont2 format

    optional arguments:
      -h, --help            show this help message and exit
      -d, --debug           debug parsing [default: False]
      -e {none,flac,ogg}, --encode {none,flac,ogg}
                            encode samples into given format [default: flac]
      -f, --force           force overwriting existing files [default: False]
      -q, --quiet           quiet operation [default: False]
      -u, --unused          show unused generators [default: True]
      --no-unused
      -o OUTPUT_DIR, --ouput-dir OUTPUT_DIR
                            output directory [default: current directory]
      -t TEMPLATE           template filename [default: empty-31.xrni]

    Convert sf2 file into renoise instrument


Use the *-o* option to specify a destination directory and *--no-unused* if you don't want to see the list of generators
which are present in the SoundFont 2 file but were not used in generating the .xnri.

*-t* allows to change the template .xnri, one is provided by default and works with renoise 3.1 at least. If you want
different default settings or generate instruments for a different version, you can provide a template of your own
and specify its filename. If the filename is not found on the filesystem, it will be looked up in the default
data patch for the python package.
If you want to provide your own template, it must follow the recommendations of the following section.

template
........

A template .xnri file will be used to have Soundfont 2 properties injected to produce a final, customized .xnri file.
It must contains at least one sample. The first sample will be used as template for all samples, others will be discarded.
The first sample must have an ADHSR modulation on volume and the filter set to a Low pass filter.
SoundFont 2 reverb and chorus effects will be mapped respectively on macro 1 and macro 2, so you may want to map those macros
on input of DSP chains to control the dry/wet amount of those effects.
You may use macro 3 and up to your liking.

compliance and limitations
..........................

The set of features found in a SoundFont 2 and .xnri doesn't fully overlap, hence this converter won't produce
a bit wise exact instrument.
First, SoundFont 2 files contains presets, which are mapped to one or more instruments. Those presets might alter how
the instruments sound, but it's a bad practice and usually only map keys range to instruments. This converter works
from SoundFont 2 instruments only.

**sf2toxrni** supports :
    * envelope volume release
    * sample panning
    * sample looping
    * sample tuning (base note, fine and coarse tuning)
    * key mapping
    * velocity mapping
    * low pass filter cutoff
    * chorus amount
    * reverb amount

**sf2toxrni** does not support :
    * vibrato
    * initial volume attenuation (always minus infinite in result, it seems like it'd be a pita to support otherwise)

sfztoxrni
---------

Similarly to the **sf2toxrni** converter, **sfztoxrni** convert SFZ files into renoise instruments (.xrni).
Renoise supporting now natively SFZ files, this converter is only useful on SFZ instruments needing
more than one modulation set (different ADHSR settings depending on the velocity/key mapping) or
a more aggressive sample filename search (like for SFZ built under case insensitive filesystem and read in
case sensitive filesystem). As SFZ support in Renoise will progress, this converter will be deprecated.

::

    usage: sfztoxrni [-h] [-d] [-e {none,flac,ogg}] [-f] [-q] [-o OUTPUT_DIR]
                     [-t TEMPLATE] [-u] [--no-unused]
                     sfz_filename [sfz_filename ...]

    GPL v3+ 2016-2017 Olivier Jolly

    positional arguments:
      sfz_filename          input file in SFZ format

    optional arguments:
      -h, --help            show this help message and exit
      -d, --debug           debug parsing [default: False]
      -e {none,flac,ogg}, --encode {none,flac,ogg}
                            encode samples into given format [default: flac]
      -f, --force           force overwriting existing files [default: False]
      -q, --quiet           quiet operation [default: False]
      -o OUTPUT_DIR, --ouput-dir OUTPUT_DIR
                            output directory [default: current directory]
      -t TEMPLATE           template filename [default: empty-31.xrni]
      -u, --unused          show unused generators [default: True]
      --no-unused

    Convert SFZ file into renoise instrument


xrnireencode
------------

**Xrnireencode** is a command line utility to reencode samples in renoise instrument (.xrni).
It can convert to **flac** or **ogg** one or more instruments given on command line.

::

    usage: xrnireencode [-h] [-d] [-e {flac,ogg}] [-q] [-o OUTPUT_DIR]
                        xrni_filename [xrni_filename ...]

    GPL v3+ 2016 Olivier Jolly

    positional arguments:
      xrni_filename         input file in XRNI format

    optional arguments:
      -h, --help            show this help message and exit
      -d, --debug           debug parsing [default: False]
      -e {flac,ogg}, --encode {flac,ogg}
                            encode samples into given format [default: flac]
      -q, --quiet           quiet operation [default: False]
      -o OUTPUT_DIR, --ouput-dir OUTPUT_DIR
                            output directory [default: current directory]

    Reencode samples in renoise instrument

xrnicomment
-----------

**xrnicomment** is a command line utility to manipulate comments in renoise instruments (.xrni).
It can read, clear, replace and append content to the comment in one or more xrni files.
It is compatible with renoise 3.0+ instruments and intercompatible with "instrument info" tool
( http://forum.renoise.com/index.php/topic/43434-new-tool-30-instrument-info/ ).

::

    usage: xrnicomment [-h] [-d] [-a] [-e] [-m MESSAGE] [-r] [-v]
                       xrni_filename [xrni_filename ...]

    GPL v3+ 2016 Olivier Jolly

    positional arguments:
      xrni_filename         input file in XRNI format

    optional arguments:
      -h, --help            show this help message and exit
      -d, --debug           debug parsing [default: False]
      -a, --append          append to comment
      -e, --edit            edit comment
      -m MESSAGE, --message MESSAGE
                            edit message content [default reads from standard
                            input]
      -r, --remove          remove comment
      -v, --view            view comment [default action]

    Display or change XRNI comments

xrnitag
-------

**xrnitag** is a command line utility to manipulate tags in renoise instruments (.xrni).
It can read, clear, remove or add tag to the tag list in one or more xrni files.
If you want to edit a tag, you can remove it and then add the new one.
**xrnitag** is compatible with renoise 3.0+ instruments and intercompatible with "instrument info" tool
( http://forum.renoise.com/index.php/topic/43434-new-tool-30-instrument-info/ ).

::

    usage: xrnitag [-h] [-d] [-a TAGS_TO_ADD] [-c] [-r TAGS_TO_REMOVE] [-v]
                   xrni_filename [xrni_filename ...]

    GPL v3+ 2016 Olivier Jolly

    positional arguments:
      xrni_filename         input file in XRNI format

    optional arguments:
      -h, --help            show this help message and exit
      -d, --debug           debug parsing [default: False]
      -a TAGS_TO_ADD, --add TAGS_TO_ADD
                            add a tag
      -c, --clear           clear all tags
      -r TAGS_TO_REMOVE, --remove TAGS_TO_REMOVE
                            remove a tag
      -v, --view            view all tags [default action]

    Display or change XRNI tags

xrniorganise
------------

**xrniorganise** is a command line utility to organise your renoise instruments (.xrni) by tags.
It creates a view of your instrument library using instrument tags, until renoise features a native tag
powered instrument library search.
Using symbolic links to create instruments views, it is only compatible with unix OSes (patches for
supporting other OSes are welcome).

Instruments which aren't tagged at all will be linked into the "untagged" directory.

Note that you can incrementally populate your library view, as **xrniorganise** will not remove existing
instruments. It also means that if you switch a tag from *old* to *new* and run **xrniorganise** again,
your instrument(s) will be linked both in the *old* and *new* directories. When you rename or remove tags,
you can use the *-c* command line argument which makes so that your destination directory is cleaned before
having any link created.

Directory cleaning attempts to be as conservative as possible (it's a recursive removal after all, it could
damage stuff if broken) by only removing symbolic links and empty directories. It implies that any regular file
you'll place inside the destination directory will abort cleaning.

Here is an example session:

.. code:: shell

    # initially, we have some untagged, unsorted xrni files

    $ find unsorted_xrni/ sorted_xrni/

    unsorted_xrni/
    unsorted_xrni/0_Arco Strings.xrni
    unsorted_xrni/6_Zip.xrni
    unsorted_xrni/0_Flute.xrni
    unsorted_xrni/8_Oboe.xrni

    sorted_xrni/

    # we tag them with xrnitag CLI (or "instrument info" tool)

    $ xrnitag -a orchestral unsorted_xrni/0_Arco\ Strings.xrni unsorted_xrni/0_Flute.xrni unsorted_xrni/8_Oboe.xrni
    $ xrnitag -a sfx unsorted_xrni/6_Zip.xrni
    $ xrnitag -a woodwind unsorted_xrni/0_Flute.xrni unsorted_xrni/8_Oboe.xrni
    $ xrnitag -a loop unsorted_xrni/6_Zip.xrni unsorted_xrni/8_Oboe.xrni unsorted_xrni/0_Arco\ Strings.xrni

    # them, we create a library view based on those tags

    $ xrniorganise -o sorted_xrni/ unsorted_xrni/*

    # now, you can see links to actual instruments based on their tags

    $ find unsorted_xrni/ sorted_xrni/

    unsorted_xrni/
    unsorted_xrni/0_Arco Strings.xrni
    unsorted_xrni/6_Zip.xrni
    unsorted_xrni/0_Flute.xrni
    unsorted_xrni/8_Oboe.xrni

    sorted_xrni/
    sorted_xrni/woodwind
    sorted_xrni/woodwind/0_Flute.xrni
    sorted_xrni/woodwind/8_Oboe.xrni
    sorted_xrni/loop
    sorted_xrni/loop/0_Arco Strings.xrni
    sorted_xrni/loop/6_Zip.xrni
    sorted_xrni/loop/8_Oboe.xrni
    sorted_xrni/sfx
    sorted_xrni/sfx/6_Zip.xrni
    sorted_xrni/orchestral
    sorted_xrni/orchestral/0_Arco Strings.xrni
    sorted_xrni/orchestral/0_Flute.xrni
    sorted_xrni/orchestral/8_Oboe.xrni

Here is the summary of all options::

    usage: xrniorganise [-h] [-d] [-c] [-n] [-r] -o OUTPUT_DIR
                        xrni_filename [xrni_filename ...]

    GPL v3+ 2016 Olivier Jolly

    positional arguments:
      xrni_filename         input file in XRNI format

    optional arguments:
      -h, --help            show this help message and exit
      -d, --debug           debug parsing [default: False]
      -c, --clean           clean destination directory before operations
      -n, --dry-run         don't actually perform filesystem operations [default:
                            False]
      -r, --recursive       recursively parse directories [default: False]
      -o OUTPUT_DIR, --ouput-dir OUTPUT_DIR
                            output directory

    Organise XRNI according to their tags




Library use
-----------

Current, only renoise instrument API is available.
A renoise instrument is represented by the **RenoiseInstrument** class and can be loaded that way::

    from rnsutils.instrument import RenoiseInstrument
    # load an instrument from an existing xnri
    inst = RenoiseInstrument('existing.xrni')
    # now, inst.root is an objectified xml tree you can access and alter
    # inst.sample_data is a mutable list of audio files content
    inst.save('new.xrni')


CHANGELOG
## [0.9.0] - 2017-02-08
### Added
- added support for midi key influence to pitch = 0% to enable drum kit like sf2 converts
- added keymap expansion when converting sf2 (to fill all key range)
- added instrument selection via index when converting sf2
- added panning overwrite to center when converting sf2
- added support for sfz not height as midi number
- added support for default_path directive
- added support for random round robin directives (lorand, hirand)

### Changed
- robustify instrument saving (creating temporary file and renaming it on success)

## [0.8.0] - 2016-02-02
### Added
- added xrnireencode CLI
- added xrnicomment CLI
- added xrnitag CLI
- added xrniorganise CLI

### Changed
- sf2toxrni and sfztoxrni default encoding is now flac


rnsutils/sfztoxrni.py
# sfztoxrni. convert SFZ file to renoise instrument
# Copyright (C) 2016  Olivier Jolly
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from __future__ import print_function

import argparse
import logging
import math
import sys

import os
import re
from copy import deepcopy

from rnsutils.instrument import RenoiseInstrument, second_to_renoise_time, db_to_renoise_volume
from rnsutils.utils import ENCODING_NONE, ENCODING_FLAC, ENCODING_OGG, encode_audio_file

__date__ = '2016-01-28'
__updated__ = '2017-02-08'
__author__ = 'olivier@pcedev.com'


def search_case_insensitive_path(path):
    if os.path.exists(path) or path == '':
        return path

    head, tail = os.path.split(path)

    existing_head = search_case_insensitive_path(head)
    if existing_head is None:
        return None
    if existing_head == '':
        existing_head = '.'

    tail = tail.lower()
    for sub in os.listdir(existing_head):
        if sub.lower() == tail:
            return os.path.join(existing_head, sub)

    return None


SFZ_NOTE_LETTER_OFFSET = {'a': 9, 'b': 11, 'c': 0, 'd': 2, 'e': 4, 'f': 5, 'g': 7}


def sfz_note_to_midi_key(sfz_note):
    try:
        return int(sfz_note)
    except ValueError:
        sharp = '#' in sfz_note
        letter = sfz_note[0].lower()
        octave = int(sfz_note[-1])
        return SFZ_NOTE_LETTER_OFFSET[letter] + ((octave + 1) * 12) + (1 if sharp else 0)


class SfzToXrni(object):
    def __init__(self, sfz_path, show_unused=False, encoding=ENCODING_NONE, **kwargs):
        self.encoding = encoding
        self.sfz_path = sfz_path
        self.show_unused = show_unused
        self.sfz_default_path = ''

    def load_default_sample_settings(self, renoise_global_sample, renoise_global_modulation_set):
        renoise_global_modulation_set.Devices.SampleMixerModulationDevice.Cutoff.Value = self.freq_to_cutoff(20000)
        renoise_global_modulation_set.Devices.SampleAhdsrModulationDevice.Attack.Value = 0
        renoise_global_modulation_set.Devices.SampleAhdsrModulationDevice.Hold.Value = 0
        renoise_global_modulation_set.Devices.SampleAhdsrModulationDevice.Decay.Value = 0
        renoise_global_modulation_set.Devices.SampleAhdsrModulationDevice.Sustain.Value = 1
        renoise_global_modulation_set.Devices.SampleAhdsrModulationDevice.Release.Value = 0
        renoise_global_modulation_set.FilterType = RenoiseInstrument.FILTER_NONE

        renoise_global_sample.LoopRelease = True
        renoise_global_sample.LoopMode = "Off"
        renoise_global_sample.LoopStart = None
        renoise_global_sample.LoopEnd = None

        renoise_global_sample.Panning = 0.5
        renoise_global_sample.Transpose = 0
        renoise_global_sample.Finetune = 0

        renoise_global_sample.Mapping.BaseNote = 60
        renoise_global_sample.Mapping.NoteStart, renoise_global_sample.Mapping.NoteEnd = (0, 119)
        renoise_global_sample.Mapping.VelocityStart, renoise_global_sample.Mapping.VelocityEnd = (0, 127)

    def convert_instrument(self, sfz_filename, renoise_instrument):

        renoise_instrument.root.find('GlobalProperties/*[Name="SF2 reverb"]').Value = 0
        renoise_instrument.root.find('GlobalProperties/*[Name="SF2 chorus"]').Value = 0
        renoise_instrument.root.SampleGenerator.KeyzoneOverlappingMode = RenoiseInstrument.OVERLAP_ALL

        renoise_instrument.comment = "Converted from SFZ {} with sfztoxrni " \
                                     "( https://gitlab.com/zeograd/rnsutils )".format(sfz_filename)

        renoise_instrument.root.GlobalProperties.MacrosVisible = False

        with open(sfz_filename, 'rt') as sfz_file:
            # convert instrument meta data
            renoise_instrument.name = os.path.basename(sfz_filename)

            # load global properties if any
            renoise_default_sample = deepcopy(renoise_instrument.sample_template)
            renoise_default_modulation_set = deepcopy(renoise_instrument.modulation_set_template)

            self.load_default_sample_settings(renoise_default_sample, renoise_default_modulation_set)

            sfz_content = self.parse_sfz(sfz_file.readlines())

            section_idx = 0

            for section_name, section_content in sfz_content:

                if section_name in ('group', 'global'):
                    self.convert_section(section_name, section_content, renoise_default_sample,
                                         renoise_default_modulation_set, renoise_default_sample,
                                         renoise_default_modulation_set, renoise_instrument)
                    continue

                # convert sample meta data in xml
                renoise_sample = deepcopy(renoise_default_sample)
                renoise_modulation_set = deepcopy(renoise_default_modulation_set)

                # force modulation set index to 0, but the sf2 modulation set remains easily enabled if needed by user
                renoise_sample.ModulationSetIndex = 0

                renoise_sample.FileName = None

                self.convert_section(section_name, section_content, renoise_sample, renoise_modulation_set,
                                     renoise_default_sample, renoise_default_modulation_set, renoise_instrument)

                if renoise_sample.FileName:
                    renoise_instrument.root.SampleGenerator.Samples.append(renoise_sample)
                    renoise_instrument.root.SampleGenerator.ModulationSets.append(renoise_modulation_set)

                    # copy wav content from sfz to renoise
                    sample_filename = search_case_insensitive_path(
                        os.path.join(self.sfz_path, self.sfz_default_path, str(renoise_sample.FileName)))

                    if sample_filename is None:
                        logging.info("missing sample file '%s'", renoise_sample.FileName)
                    else:
                        with open(sample_filename, 'rb') as sample_content:
                            renoise_instrument.sample_data.append(encode_audio_file(sample_content.read(), self.encoding))

                section_idx += 1

    def parse_sfz(self, sfz):

        match_section = re.compile('^<(.*)>$')

        sections = []
        partial_section = {}
        current_section = None

        for line in sfz:
            line = line.strip()
            matching = match_section.match(line)
            if matching:
                if partial_section:
                    sections.append((current_section, partial_section))
                    partial_section = {}
                current_section = matching.group(1)
            elif "=" in line:
                key, _, value = line.partition('=')
                partial_section[key] = value

        if partial_section:
            sections.append((current_section, partial_section))
            partial_section = {}

        return sections

    def freq_to_cutoff(self, param):
        return 127. * max(0, min(1, math.log(param / 130.) / 5)) if param else None

    def to_renoise_time(self, envelope_attenuation):
        return math.pow(envelope_attenuation / 60., 1 / 3.) if envelope_attenuation else None

    def convert_section(self, section_name, section_content, renoise_sample, renoise_modulation_set,
                        renoise_default_sample, renoise_default_modulation_set, renoise_instrument):

        unused_keys = []

        for key in section_content.keys():
            value = section_content[key].lower()
            if key == 'sample':
                renoise_sample.FileName = re.sub(r'\\+', r'/', value)

                # remove first char if file separator
                if renoise_sample.FileName.pyval[0] in (r'/', r'\\'):
                    renoise_sample.FileName = renoise_sample.FileName.pyval[1:]

                renoise_sample.Name, _ = os.path.splitext(os.path.basename(renoise_sample.FileName.pyval))
            elif key == 'lokey':
                renoise_sample.Mapping.NoteStart = sfz_note_to_midi_key(value)
            elif key == 'hikey':
                renoise_sample.Mapping.NoteEnd = sfz_note_to_midi_key(value)
            elif key == 'lovel':
                renoise_sample.Mapping.VelocityStart = value
            elif key == 'hivel':
                renoise_sample.Mapping.VelocityEnd = value
            elif key == 'pitch_keycenter':
                renoise_sample.Mapping.BaseNote = sfz_note_to_midi_key(value)
            elif key == 'ampeg_attack':
                renoise_modulation_set.Devices.SampleAhdsrModulationDevice.Attack.Value = second_to_renoise_time(
                    float(value))
            elif key == 'ampeg_hold':
                renoise_modulation_set.Devices.SampleAhdsrModulationDevice.Hold.Value = second_to_renoise_time(
                    float(value))
            elif key == 'ampeg_decay':
                renoise_modulation_set.Devices.SampleAhdsrModulationDevice.Decay.Value = second_to_renoise_time(
                    float(value))
            elif key == 'ampeg_sustain':
                renoise_modulation_set.Devices.SampleAhdsrModulationDevice.Sustain.Value = 1 - float(value) / 100.
            elif key == 'ampeg_release':
                renoise_modulation_set.Devices.SampleAhdsrModulationDevice.Release.Value = second_to_renoise_time(
                    float(value))
            elif key == 'tune':
                renoise_sample.Finetune = int(128 * float(value) / 100.)
            elif key == 'transpose':
                renoise_sample.Transpose = value
            elif key == 'volume':
                renoise_sample.Volume = db_to_renoise_volume(float(value))
            elif key == 'pan':
                renoise_sample.Panning = 0.5 + float(value) / 200.
            elif key == 'fil_type':
                if value.lower().startswith('lp'):
                    renoise_modulation_set.FilterType = RenoiseInstrument.FILTER_CLEAN_LP
                elif value.lower().startswith('hp'):
                    renoise_modulation_set.FilterType = RenoiseInstrument.FILTER_CLEAN_HP
                else:
                    renoise_modulation_set.FilterType = RenoiseInstrument.FILTER_NONE
            elif key == 'loop_mode':
                if value == 'one_shot':
                    renoise_sample.LoopMode = RenoiseInstrument.LOOP_FORWARD
                else:
                    renoise_sample.LoopMode = RenoiseInstrument.LOOP_NONE
            elif key == 'loop_start':
                renoise_sample.LoopStart = value
            elif key == 'loop_end':
                renoise_sample.LoopEnd = value
            elif key == 'seq_position':
                if int(value) > 1:
                    logging.info("Switch the entire instrument to sample cycling")
                    renoise_instrument.root.SampleGenerator.KeyzoneOverlappingMode = RenoiseInstrument.OVERLAP_CYCLE
            elif key == 'seq_length':
                pass
            elif key == 'lorand':
                if float(value) > 0:
                    logging.info("Switch the entire instrument to random sample round robin")
                    renoise_instrument.root.SampleGenerator.KeyzoneOverlappingMode = RenoiseInstrument.OVERLAP_RANDOM
            elif key == 'hirand':
                if float(value) < 1:
                    logging.info("Switch the entire instrument to random sample round robin")
                    renoise_instrument.root.SampleGenerator.KeyzoneOverlappingMode = RenoiseInstrument.OVERLAP_RANDOM
            elif key == 'group':
                logging.info("Ignoring mute group info")
            elif key == 'off_by':
                logging.info("Ignoring mute group info")
            elif key == 'default_path':
                self.sfz_default_path = re.sub(r'\\+', r'/', value)

                # remove first char if file separator
                if self.sfz_default_path[0] in (r'/', r'\\'):
                    self.sfz_default_path = self.sfz_default_path[1:]

            else:
                unused_keys.append(key)

        # disable loop when no loop duration was set
        if renoise_sample.LoopStart.pyval is None or renoise_sample.LoopEnd.pyval is None:
            renoise_sample.LoopMode = RenoiseInstrument.LOOP_NONE

        if unused_keys and self.show_unused:
            sys.stderr.write(
                "Unused key(s) for section {}:\n{}\n".format(section_name,
                                                             "\n".join([" - " + k for k in unused_keys])))


def main(argv=None):
    program_name = os.path.basename(sys.argv[0])
    program_version = "v0.9"
    program_build_date = "%s" % __updated__

    program_version_string = 'sfztoxrni %s (%s)' % (program_version, program_build_date)
    program_longdesc = '''Convert SFZ file into renoise instrument'''
    program_license = "GPL v3+ 2016-2017 Olivier Jolly"

    if argv is None:
        argv = sys.argv[1:]

    try:
        parser = argparse.ArgumentParser(epilog=program_longdesc,
                                         description=program_license)
        parser.add_argument("-d", "--debug", dest="debug", action="store_true",
                            default=False,
                            help="debug parsing [default: %(default)s]")
        parser.add_argument("-e", "--encode", dest="encoding", choices=[ENCODING_NONE, ENCODING_FLAC, ENCODING_OGG],
                            default=ENCODING_FLAC, help="encode samples into given format [default: %(default)s]")
        parser.add_argument("-f", "--force", dest="force", default=False, action="store_true",
                            help="force overwriting existing files [default: %(default)s]")
        parser.add_argument("-q", "--quiet", dest="quiet", action="store_true", default=False,
                            help="quiet operation [default: %(default)s]")
        parser.add_argument("-o", "--ouput-dir", dest="output_dir",
                            help="output directory [default: current directory]")
        parser.add_argument("-t", dest="template", help="template filename [default: %(default)s]",
                            default="empty-31.xrni")
        parser.add_argument("-u", "--unused", dest="show_unused", action="store_true", default=True,
                            help="show unused generators [default: %(default)s]")
        parser.add_argument("--no-unused", dest="show_unused", action="store_false")
        parser.add_argument("-v", "--version", action="version", version=program_version_string)

        parser.add_argument("sfz_filename", help="input file in SFZ format", nargs="+")

        # process options
        opts = parser.parse_args(argv)

    except Exception as e:
        indent = len(program_name) * " "
        sys.stderr.write(program_name + ": " + repr(e) + "\n")
        sys.stderr.write(indent + "  for help use --help")
        return 2

    if opts.debug:
        logging.root.setLevel(logging.DEBUG)
    else:
        logging.root.setLevel(logging.INFO)

    for sfz_filename in opts.sfz_filename:

        if not opts.quiet:
            print("Reading instrument from '{}'".format(sfz_filename))

        # noinspection PyBroadException
        try:
            sfz_path = os.path.dirname(sfz_filename)
            sfz_to_xrni = SfzToXrni(sfz_path=sfz_path, **vars(opts))

            renoise_instrument = RenoiseInstrument(template_filename=opts.template)
            sfz_to_xrni.convert_instrument(sfz_filename, renoise_instrument)

            filename_without_extension, extension = os.path.splitext(os.path.basename(sfz_filename))
            output_filename = os.path.join(opts.output_dir or sfz_path, '{}.xrni'.format(filename_without_extension))
            renoise_instrument.save(output_filename, overwrite=opts.force)

            if not opts.quiet:
                print("Saved {}".format(output_filename))
        except Exception:
            if not opts.quiet:
                print("FAILED")
            logging.exception("Failed to convert instrument")

    return 0


if __name__ == "__main__":
    sys.exit(main())
